nmt.go

```diff
@@ -0,0 +1,268 @@
+}
+
+type NamespacedMerkleTree struct {
+	nidLen     int
+	baseHasher TreeHasher
+	tree       *merkletree.Tree
+
+	// just cache stuff until we pass in a store and keep all nodes in there
+	leafs      []NamespacePrefixedData
+	leafHashes [][]byte
+	// this can be used to efficiently lookup the range for an
+	// existing namespace without iterating through the leafs
+	namespaceRanges map[string]merkletree.LeafRange
+}
+
+func (n NamespacedMerkleTree) Prove(index int) (root []byte, rawProof [][]byte, proofIdx int, totalNumLeafs int, err error) {
+	subTreeHasher := internal.NewCachedSubtreeHasher(n.leafHashes, n.baseHasher)
+	// XXX do we really need to return the root?
+	_, _, root = n.Root()
+	proofIdx = index
+	totalNumLeafs = len(n.leafs)
+	// TODO: store nodes and re-use the hashes instead recomputing parts of the tree here
+	rawProof, err = merkletree.BuildRangeProof(index, index+1, subTreeHasher)
+
+	return
+}
+
+func (n NamespacedMerkleTree) ProveNamespace(nID NamespaceID) (proofStart int, proofEnd int, proof [][]byte, foundLeafs []NamespacePrefixedData, leafHashes [][]byte) {
+	found, proofStart, proofEnd := n.foundNamespaceID(nID)
+
+	// If we did not find the namespace, that either means that there is a gap in the tree
+	// i.e. a range where this namespace would live if it was pushed,
+	// or, it is smaller or greater than any pushed namespace.
+	// XXX this can probably be simplified, too:
+	foundRangeStart := false
+	if !found {
+		// Generate a proof for an absence using the
+		// range the namespace would be in
+		// TODO: document this as a proper godoc comment
+		var prevLeaf NamespacePrefixedData
+		for index, curLeaf := range n.leafs {
+			if index == 0 {
+				continue
+			}
+			prevNs := prevLeaf.NamespaceID()
+			currentNs := curLeaf.NamespaceID()
+			// Note that here we would also care for the case
+			// current < nId < prevNs
+			// but we only allow pushing leafs with ascending namespaces;
+			// i.e. prevNs <= currentNs is always true.
+			// Also we only check for strictly smaller: prev < nid < current
+			// because if we either side was equal, we would have found the
+			// namespace before.
+			if prevNs.Less(nID) && nID.Less(currentNs) {
+				if !foundRangeStart {
+					foundRangeStart = true
+					proofStart = index
+				}
+				proofEnd = index + 1
+			}
+			prevLeaf = curLeaf
+		}
+	}
+
+	if found || foundRangeStart {
+		subTreeHasher := internal.NewCachedSubtreeHasher(n.leafHashes, n.baseHasher)
+		var err error
+		proof, err = merkletree.BuildRangeProof(proofStart, proofEnd, subTreeHasher)
+		if err != nil {
+			panic(fmt.Sprintf("unexpected err: %v, range: [%v, %v)", err, proofStart, proofEnd))
+		}
+	}
+	proofMessages := n.leafs[proofStart:proofEnd]
+	if found {
+		return proofStart, proofEnd, proof, proofMessages, nil
+	}
+	// XXX further clarify the edge-case (!found && !foundRangeStart)
+	// !found: we return leafHashes instead:
+	return proofStart, proofEnd, proof, nil, n.leafHashes[proofStart:proofEnd]
+}
```

Comment:
Something I wondered about: If the namespace to prove `nID` can't be found in the tree and `nID < minNS` or `nID > maxNS` (where `minNs, maxNs,_ := tree.Root()`), shouldn't the prove of exclusion be either an inclusion proof of the left- or right-most leaf?

Currently, in the [prototype code](This should be the same logic as https://github.com/lazyledger/lazyledger-prototype/blob/2aeca6f55ad389b9d68034a0a7038f80a8d2982e/simpleblock.go#L105-L160
) as well as above application proofs return:
`(proofStart, proofEnd, proof, messages, hashes) == (0, 0, nil, nil, nil)` if `nID < minNS || maxNS < nID`.

Reply:
Capturing the discussing with @adlerjohn and @musalbas:
 - in the case the that  min and max namespace of the root is not within range of the nid, no proof is required
 - the root already signifies that this nid does not exist in that particular tree
 - retuning a sentinel error (signifying and as before returning no proof) probably makes the most sense here